### ISSUE TITLE
user_external backend missing when revalidating password in checkPasswordNoLogging

### ISSUE THREAD

**Issue by neffets** [nextcloud/server#7809]
I'm experiencing this problem with user_external "IMAP" authentication.
Using: nextcloud 12.0.4.3

### Steps to reproduce

pre-1: I have some users with internal password and some using user_external OC_User_IMAP
pre-2: the users with internal password have no problem
pre-3: the imap-user has in db-table: "oc_users" no password set (but has an entry, with empty password), and an entry in the "oc_users_external"

1. Login with imap-user: logging enabled: imap ok, password ok, session created ok, session-cookie transmitted ok
2. Redirect to /index.php/apps/files, there checking fails and session will be invalidated

For DEBUG results of me, see at end of this comment

### Expected behaviour
After first successful login, the rest should work too

### Actual behaviour
the second request  invalidateSession()

### Server configuration

**Operating system**:
Debian 8 (jessie)

**Web server:**
apache-2.4.10

**Database:**
mysql 5.7

**PHP version:**
7.1.12

**Nextcloud version:** (see Nextcloud admin page)
12.0.4.3 

**Updated from an older Nextcloud/ownCloud or fresh install:**
former owncloud till version 9, the nextcloud 10, 11, 12

**Where did you install Nextcloud from:**
Docker "nextcloud:12.0.4-apache"  with further:
>     docker-php-ext-configure imap --with-kerberos --with-imap-ssl && \
>     docker-php-ext-install imap; \
to enable imap support

**Signing status:**
<details>
<summary>Signing status</summary>

```
occ integrity:check-core

- EXTRA_FILE:
    - lib/private/User/.Manager.php.swp:
      - expected: 
      - current: e8b1aeb1f74f729e0453d191767e9df20b5a136dba4861c90e5261afab19742d76984dc1f04ff1280e1780b8f83eb8cbc96741198c22919d5294950ad88ff803

```
</details>

**List of activated apps:**
<details>
<summary>App list</summary>

```
Enabled:
  - activity: 2.5.2
  - calendar: 1.5.7
  - comments: 1.2.0
  - contacts: 2.0.1
  - dav: 1.3.0
  - federatedfilesharing: 1.2.0
  - files: 1.7.2
  - files_external: 1.3.0
  - files_sharing: 1.4.0
  - files_trashbin: 1.2.0
  - files_versions: 1.5.0
  - firstrunwizard: 2.1
  - gallery: 17.0.0
  - logreader: 2.0.0
  - lookup_server_connector: 1.0.0
  - nextcloud_announcements: 1.1
  - notifications: 2.0.0
  - oauth2: 1.0.5
  - password_policy: 1.2.2
  - provisioning_api: 1.2.0
  - serverinfo: 1.2.0
  - sharebymail: 1.2.0
  - systemtags: 1.2.0
  - theming: 1.3.0
  - twofactor_backupcodes: 1.1.1
  - updatenotification: 1.2.0
  - user_external: 0.4
  - workflowengine: 1.2.0
Disabled:
  - admin_audit
  - bruteforcesettings
  - encryption
  - federation
  - files_pdfviewer
  - files_texteditor
  - files_videoplayer
  - survey_client
  - user_ldap
```
</details>

**Nextcloud configuration:**
<details>
<summary>Config report</summary>

```
{
    "system": {
        "instanceid": "XXX92kb4g6tx",
        "passwordsalt": "***REMOVED SENSITIVE VALUE***",
        "secret": "***REMOVED SENSITIVE VALUE***",
        "trusted_domains": [
            "localhost",
            "own.neffets.de"
        ],
        "datadirectory": "\/var\/www\/html\/data",
        "overwrite.cli.url": "http:\/\/localhost\/",
        "version": "12.0.4.3",
        "installed": true,
        "installedat": "1337464107.6398",
        "lastupdatedat": "1337805530.2802",
        "dbtype": "mysql",
        "dbname": "nextcloud",
        "dbhost": "mysql_db:3306",
        "dbport": "",
        "dbtableprefix": "oc_",
        "dbuser": "***REMOVED SENSITIVE VALUE***",
        "dbpassword": "***REMOVED SENSITIVE VALUE***",
        "mysql.utf8mb4": true,
        "logtimezone": "UTC",
        "blacklisted_files": [
            ".htaccess"
        ],
        "mail_domain": "neffets.de",
        "loglevel": "0",
        "defaultapp": "files",
        "knowledgebaseenabled": true,
        "knowledgebaseurl": "http:\/\/api.apps.owncloud.com\/v1",
        "appstoreenabled": true,
        "appstoreurl": "http:\/\/api.apps.owncloud.com\/v1",
        "updatechecker": "false",
        "maintenance": false,
        "forcessl": true,
        "theme": "",
        "mail_from_address": "owncloud",
        "mail_smtpmode": "php",
        "user_backends": [
            {
                "class": "OC_User_IMAP",
                "arguments": [
                    "{imap.neffets.de:993\/imap\/ssl}INBOX"
                ]
            }
        ],
        "debug": true,
        "trashbin_retention_obligation": "auto",
        "filelocking.enabled": "false",
        "default_language": "de",
        "redis": {
            "host": "nextcloud_redis",
            "port": 6379
        },
        "memcache.distributed": "\\OC\\Memcache\\Redis",
        "memcache.locking": "\\OC\\Memcache\\Redis",
        "htaccess.RewriteBase": "\/",
        "memcache.local": "\\OC\\Memcache\\APCu",
        "apps_paths": [
            {
                "path": "\/var\/www\/html\/apps",
                "url": "\/apps",
                "writable": false
            },
            {
                "path": "\/var\/www\/html\/custom_apps",
                "url": "\/custom_apps",
                "writable": true
            }
        ]
    }
}
```
</details>

**Are you using external storage, if yes which one:** local/smb/sftp/...
NO

**Are you using encryption:** yes/no
NO

**Are you using an external user-backend, if yes which one:** LDAP/ActiveDirectory/Webdav/...
YES IMAP

Imap backend works.

### Client configuration
**Browser:**
any, here firefox-57

**Operating system:**
any, here ubuntu-17.10

#### Web server error log
<details>
<summary>Web server error log</summary>

```
nextcloud_www.1.o1kqlmlcknus@neffets.de    | [Fri Jan 12 09:27:15.235410 2018] [php7:notice] [pid 96] [client 10.0.10.6:56140] CHK_BACK  OC\\User\\Database
nextcloud_www.1.o1kqlmlcknus@neffets.de    | [Fri Jan 12 09:27:15.236037 2018] [php7:notice] [pid 96] [client 10.0.10.6:56140] CHK_BACK  OC\\User\\Database uid=
nextcloud_www.1.o1kqlmlcknus@neffets.de    | [Fri Jan 12 09:27:15.236127 2018] [php7:notice] [pid 96] [client 10.0.10.6:56140] CHK_BACK  OC_User_IMAP
nextcloud_www.1.o1kqlmlcknus@neffets.de    | [Fri Jan 12 09:27:15.335779 2018] [php7:notice] [pid 96] [client 10.0.10.6:56140] CHK_BACK  OC_User_IMAP uid=neffets
nextcloud_www.1.o1kqlmlcknus@neffets.de    | 10.0.10.6 - - [12/Jan/2018:09:26:54 +0000] "POST /index.php/heartbeat HTTP/1.1" 200 832 "-" "Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:57.0) Gecko/20100101 Firefox/57.0"
nextcloud_www.1.o1kqlmlcknus@neffets.de    | 10.0.10.6 - - [12/Jan/2018:09:26:49 +0000] "POST /index.php/login?redirect_url=/index.php/apps/files/ HTTP/1.1" 303 962 "-" "Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:57.0) Gecko/20100101 Firefox/57.0"
nextcloud_www.1.o1kqlmlcknus@neffets.de    | [Fri Jan 12 09:27:15.583987 2018] [php7:notice] [pid 66] [client 10.0.10.6:56208] SESSION wir haben ID token=7771b30a9cc9d1282f26131d49a2ac62
nextcloud_www.1.o1kqlmlcknus@neffets.de    | [Fri Jan 12 09:27:15.584819 2018] [php7:notice] [pid 66] [client 10.0.10.6:56208] CHECK_tokenCred lastCheck=0 >? now-60*5 = 1515748935
nextcloud_www.1.o1kqlmlcknus@neffets.de    | [Fri Jan 12 09:27:15.586079 2018] [php7:notice] [pid 66] [client 10.0.10.6:56208] OC\\User\\Manager  user=neffets  pwd=bol6dom6,STS
nextcloud_www.1.o1kqlmlcknus@neffets.de    | [Fri Jan 12 09:27:15.586364 2018] [php7:notice] [pid 66] [client 10.0.10.6:56208] CHK_BACK  OC\\User\\Database
nextcloud_www.1.o1kqlmlcknus@neffets.de    | [Fri Jan 12 09:27:15.586898 2018] [php7:notice] [pid 66] [client 10.0.10.6:56208] CHK_BACK  OC\\User\\Database uid=
nextcloud_www.1.o1kqlmlcknus@neffets.de    | [Fri Jan 12 09:27:15.587018 2018] [php7:notice] [pid 66] [client 10.0.10.6:56208] CHECK_PWD = checkpwd (1 / 0)
nextcloud_www.1.o1kqlmlcknus@neffets.de    | [Fri Jan 12 09:27:15.587113 2018] [php7:notice] [pid 66] [client 10.0.10.6:56208] CHK_BACK  OC\\User\\Database
nextcloud_www.1.o1kqlmlcknus@neffets.de    | [Fri Jan 12 09:27:15.587475 2018] [php7:notice] [pid 66] [client 10.0.10.6:56208] CHK_BACK  OC\\User\\Database uid=
nextcloud_www.1.o1kqlmlcknus@neffets.de    | [Fri Jan 12 09:27:15.634430 2018] [php7:notice] [pid 66] [client 10.0.10.6:56208] SESSION invalidate Session
nextcloud_www.1.o1kqlmlcknus@neffets.de    | 10.0.10.6 - - [12/Jan/2018:09:27:15 +0000] "GET /index.php/apps/files/ HTTP/1.1" 303 1598 "-" "Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:57.0) Gecko/20100101 Firefox/57.0"
nextcloud_www.1.o1kqlmlcknus@neffets.de    | 10.0.10.6 - - [12/Jan/2018:09:27:15 +0000] "GET /index.php/login?redirect_url=/index.php/apps/files/ HTTP/1.1" 200 4329 "-" "Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:57.0) Gecko/20100101 Firefox/57.0"

```
</details>

#### Nextcloud log (data/nextcloud.log)
<details>
<summary>Nextcloud log</summary>

```
{"reqId":"s3EXF1UtmOx5rb7PjHWh","level":0,"time":"2018-01-12T09:07:49+00:00","remoteAddr":"10.0.10.6","user":"--","app":"core","method":"GET","url":"\/index.php\/login?redirect_url=\/index.php\/apps\/files\/","message":"Scss is disabled for \/var\/www\/html\/core\/css\/sh
are.scss, ignoring","userAgent":"Mozilla\/5.0 (X11; Ubuntu; Linux x86_64; rv:57.0) Gecko\/20100101 Firefox\/57.0","version":"12.0.4.3"}
{"reqId":"5jqe148aiSkB29BH7dQg","level":2,"time":"2018-01-12T09:09:00+00:00","remoteAddr":"10.0.10.6","user":"neffets","app":"core","method":"GET","url":"\/index.php\/apps\/files\/","message":"Login failed: 'neffets' (Remote IP: '10.0.10.6')","userAgent":"Mozilla\/5.0 (X1
1; Ubuntu; Linux x86_64; rv:57.0) Gecko\/20100101 Firefox\/57.0","version":"12.0.4.3"}

{"reqId":"5jqe148aiSkB29BH7dQg","level":0,"time":"2018-01-12T09:09:00+00:00","remoteAddr":"10.0.10.6","user":"--","app":"PHP","method":"GET","url":"\/index.php\/apps\/files\/","message":"A session had already been started - ignoring session_start() at \/var\/www\/html\/li
b\/private\/Session\/Internal.php#172","userAgent":"Mozilla\/5.0 (X11; Ubuntu; Linux x86_64; rv:57.0) Gecko\/20100101 Firefox\/57.0","version":"12.0.4.3"}
{"reqId":"5jqe148aiSkB29BH7dQg","level":0,"time":"2018-01-12T09:09:00+00:00","remoteAddr":"10.0.10.6","user":"--","app":"no app in context","method":"GET","url":"\/index.php\/apps\/files\/","message":"Current user is not logged in","userAgent":"Mozilla\/5.0 (X11; Ubuntu; 
Linux x86_64; rv:57.0) Gecko\/20100101 Firefox\/57.0","version":"12.0.4.3"}
{"reqId":"S2axmNwxW2fX8k7qbHRS","level":0,"time":"2018-01-12T09:09:00+00:00","remoteAddr":"10.0.10.6","user":"--","app":"core","method":"GET","url":"\/index.php\/login?redirect_url=\/index.php\/apps\/files\/","message":"Scss is disabled for \/var\/www\/html\/core\/css\/se
rver.scss, ignoring","userAgent":"Mozilla\/5.0 (X11; Ubuntu; Linux x86_64; rv:57.0) Gecko\/20100101 Firefox\/57.0","version":"12.0.4.3"}
{"reqId":"S2axmNwxW2fX8k7qbHRS","level":0,"time":"2018-01-12T09:09:00+00:00","remoteAddr":"10.0.10.6","user":"--","app":"core","method":"GET","url":"\/index.php\/login?redirect_url=\/index.php\/apps\/files\/","message":"Scss is disabled for \/var\/www\/html\/core\/css\/sh
are.scss, ignoring","userAgent":"Mozilla\/5.0 (X11; Ubuntu; Linux x86_64; rv:57.0) Gecko\/20100101 Firefox\/57.0","version":"12.0.4.3"}
{"reqId":"jB0D0A4e7YOcQtgX20WA","level":0,"time":"2018-01-12T09:13:00+00:00","remoteAddr":"10.0.10.6","user":"--","app":"PHP","method":"POST","url":"\/index.php\/login?redirect_url=\/index.php\/apps\/files\/","message":"error_log() expects parameter 2 to be integer, strin
g given at \/var\/www\/html\/apps\/user_external\/lib\/imap.php#68","userAgent":"Mozilla\/5.0 (X11; Ubuntu; Linux x86_64; rv:57.0) Gecko\/20100101 Firefox\/57.0","version":"12.0.4.3"}
{"reqId":"kNZxPuHHXMgT4t2rCGO0","level":2,"time":"2018-01-12T09:13:01+00:00","remoteAddr":"10.0.10.6","user":"neffets","app":"core","method":"GET","url":"\/index.php\/apps\/files\/","message":"Login failed: 'neffets' (Remote IP: '10.0.10.6')","userAgent":"Mozilla\/5.0 (X1
1; Ubuntu; Linux x86_64; rv:57.0) Gecko\/20100101 Firefox\/57.0","version":"12.0.4.3"}
{"reqId":"kNZxPuHHXMgT4t2rCGO0","level":0,"time":"2018-01-12T09:13:01+00:00","remoteAddr":"10.0.10.6","user":"--","app":"PHP","method":"GET","url":"\/index.php\/apps\/files\/","message":"A session had already been started - ignoring session_start() at \/var\/www\/html\/li
b\/private\/Session\/Internal.php#172","userAgent":"Mozilla\/5.0 (X11; Ubuntu; Linux x86_64; rv:57.0) Gecko\/20100101 Firefox\/57.0","version":"12.0.4.3"}

```
</details>

I tracked it down to the    lib/private/User/Manager.php:   about line 200 in function checkPasswordNoLogging

I added some debug line to see, which backend is asked and what the response is:

(1) the login-Request shows:
```
nextcloud_www.1.o1kqlmlcknus@neffets.de    | [Fri Jan 12 09:27:15.235410 2018] [php7:notice] [pid 96] [client 10.0.10.6:56140] CHK_BACK  OC\\User\\Database
nextcloud_www.1.o1kqlmlcknus@neffets.de    | [Fri Jan 12 09:27:15.236037 2018] [php7:notice] [pid 96] [client 10.0.10.6:56140] CHK_BACK  OC\\User\\Database uid=
nextcloud_www.1.o1kqlmlcknus@neffets.de    | [Fri Jan 12 09:27:15.236127 2018] [php7:notice] [pid 96] [client 10.0.10.6:56140] CHK_BACK  OC_User_IMAP
nextcloud_www.1.o1kqlmlcknus@neffets.de    | [Fri Jan 12 09:27:15.335779 2018] [php7:notice] [pid 96] [client 10.0.10.6:56140] CHK_BACK  OC_User_IMAP uid=neffets
```
so the internal AND the imap-backend are checked, result is "neffets" => OK

(2) next request /index.php/apps/files FAILS because:
```
nextcloud_www.1.o1kqlmlcknus@neffets.de    | 10.0.10.6 - - [12/Jan/2018:09:26:49 +0000] "POST /index.php/login?redirect_url=/index.php/apps/files/ HTTP/1.1" 303 962 "-" "Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:57.0) Gecko/20100101 Firefox/57.0"
nextcloud_www.1.o1kqlmlcknus@neffets.de    | [Fri Jan 12 09:27:15.583987 2018] [php7:notice] [pid 66] [client 10.0.10.6:56208] SESSION wir haben ID token=7771b30a9cc9d1282f26131d49a2ac62
nextcloud_www.1.o1kqlmlcknus@neffets.de    | [Fri Jan 12 09:27:15.584819 2018] [php7:notice] [pid 66] [client 10.0.10.6:56208] CHECK_tokenCred lastCheck=0 >? now-60*5 = 1515748935
nextcloud_www.1.o1kqlmlcknus@neffets.de    | [Fri Jan 12 09:27:15.586079 2018] [php7:notice] [pid 66] [client 10.0.10.6:56208] OC\\User\\Manager  user=neffets  pwd=**MASKED**
nextcloud_www.1.o1kqlmlcknus@neffets.de    | [Fri Jan 12 09:27:15.586364 2018] [php7:notice] [pid 66] [client 10.0.10.6:56208] CHK_BACK  OC\\User\\Database
nextcloud_www.1.o1kqlmlcknus@neffets.de    | [Fri Jan 12 09:27:15.586898 2018] [php7:notice] [pid 66] [client 10.0.10.6:56208] CHK_BACK  OC\\User\\Database uid=
nextcloud_www.1.o1kqlmlcknus@neffets.de    | [Fri Jan 12 09:27:15.587018 2018] [php7:notice] [pid 66] [client 10.0.10.6:56208] CHECK_PWD = checkpwd (1 / 0)
nextcloud_www.1.o1kqlmlcknus@neffets.de    | [Fri Jan 12 09:27:15.587113 2018] [php7:notice] [pid 66] [client 10.0.10.6:56208] CHK_BACK  OC\\User\\Database
nextcloud_www.1.o1kqlmlcknus@neffets.de    | [Fri Jan 12 09:27:15.587475 2018] [php7:notice] [pid 66] [client 10.0.10.6:56208] CHK_BACK  OC\\User\\Database uid=
nextcloud_www.1.o1kqlmlcknus@neffets.de    | [Fri Jan 12 09:27:15.634430 2018] [php7:notice] [pid 66] [client 10.0.10.6:56208] SESSION invalidate Session
nextcloud_www.1.o1kqlmlcknus@neffets.de    | 10.0.10.6 - - [12/Jan/2018:09:27:15 +0000] "GET /index.php/apps/files/ HTTP/1.1" 303 1598 "-" "Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:57.0) Gecko/20100101 Firefox/57.0"
```
**Here ONLY the internal backend is available!**

I dont know why at this moment, the imap-user-external backend is not listed and not checked => it fails!

**Comment by blizzz** [comment]
> the imap-user has in db-table: "oc_users" no password set (but has an entry, with empty password), and an entry in the "oc_users_external"

@icewind1991 sounds bogus, isn't it? I'd expect to have it just one entry in oc_users_external

**Comment by tisoft** [comment]
I had the same behavior.

In my case I even had a password set in the users table and coincidentally it was the same as the IMAP password, but never the less, I got logged out after a few seconds every time...

The reason for the entry in my case was, that I first installed nextcloud with local passwords and only afterwards connected it with IMAP. So for anyone coming here, with the same situation, deleting the accounts from the users table is the fix.
